proposals/0279-multiple-trailing-closures.md

```diff
@@ -1,422 +1,383 @@
-func toggle() {
-  UIView.animate(withDuration: 1) {
-    animations: {
-      self.myView.backgroundColor = UIColor.green
-      self.myView.frame.size.width += 50
-      self.myView.frame.size.height += 20
-      self.myView.center.x += 20
+  ```swift
+  ipAddressPublisher
+    .sink { identity in
+      self.hostnames.insert(identity.hostname!)
     }
-    completion: { _ in
-      UIView.animate(withDuration: 1, delay: 0.25, options: [.autoreverse, .repeat]) {
-        animations: {
-          self.myView.frame.origin.y -= 20
-        }
-      }
+
+  ipAddressPublisher
+    .sink(receiveCompletion: { completion in
+      // handle error
+    }) { identity in
+      self.hostnames.insert(identity.hostname!)
     }
-  }
-}
-```
+  ```
 
-Here, we've eliminated the inconsistencies where some "calls" close with a single `}` and others
-with a `})`. The new syntax composes well with SwiftUI's `Button`:
+  ... but could be re-worked in light of multiple trailing closure syntax:
 
-```swift
-Button {
-  action: {
-    ...
-    ...
-    ...
-  }
+  ```swift
+  ipAddressPublisher
+    .sink { identity in
+      self.hostnames.insert(identity.hostname!)
+    }
+
+  ipAddressPublisher
+    .sink { identity in
+      self.hostnames.insert(identity.hostname!)
+    } receiveCompletion: { completion in
+      // handle error
+    }
+  ```
 
-  label: {
-    Text("Hello!")
+3. Consider SwiftUI's [Section](https://developer.apple.com/documentation/swiftui/section) view, which today avoids using `@ViewBuilder` closures for its optional header and footer:
+
+  ```swift
+  Section {
+    // content
   }
-}
-```
+  Section(header: ...) {
+    // content
+  }
+  Section(footer: ...) {
+    // content
+  }
+  Section(
+    header: ...,
+    footer: ...
+  ) {
+    // content
+  }
+  ```
 
-and our `when(then:else:)` control-flow construct:
+  ... but could be re-worked in light of multiple trailing closure syntax:
 
-```swift
-when(2 < 3) {
-  then: {
+  ```swift
+  Section {
+    // content
+  }
+  Section {
+    // content
+  } header: {
     ...
+  }
+  Section {
+    // content
+  } footer: {
     ...
   }
-  else: {
+  Section {
+    // content
+  } header: {
     ...
+  } footer: {
     ...
   }
-}
+  ```
+
+When using multiple trailing closure syntax, these APIs are all [clear at the point of use](https://swift.org/documentation/api-design-guidelines/#fundamentals), without the need to label the first trailing closure.
+
+If labelling the first trailing closure were allowed, users would have to evaluate whether to include the label on a case by case basis, which would inevitably lead to linter and style guide rules to prohibit it. So, in conjunction with the new syntax rules, we propose an amendment to the [API Design Guidelines](https://swift.org/documentation/api-design-guidelines):
+
+> Name functions assuming that the argument label of the first trailing closure will be dropped. Include meaningful argument labels for all subsequent trailing closures.
```

Comment:
Can you add a discussion of how existing standard library API that doesn't follow this convention will be handled?  `Sequence.first(where:)` is a good example of this.

Reply:
Pinging @airspeedswift for that part.

Reply:
`drop(while:)` is probably the urbug in the std lib.

Reply:
But going back and fixing existing issues (as opposed to setting a new standard) should be left to a separate discussion rather than mixed into this one.

Reply:
Applying this advice would allow us to amend and reintroduce `countWhere` as well.

Reply:
Right, but I think the proposal should make it clear that this is a viable future direction if that’s the case.  One sentence would be enough.

Reply:
IMO it would be out of place to suggest only-tangentially-related proposals in this proposal text.